sdks/java/core/src/main/java/org/apache/beam/sdk/testing/BigqueryMatcher.java

```diff
@@ -151,17 +151,24 @@ QueryResponse queryWithRetries(Bigquery bigqueryClient, QueryRequest queryConten
     IOException lastException = null;
     do {
       try {
-        return bigqueryClient.jobs().query(projectId, queryContent).execute();
+        response = bigqueryClient.jobs().query(projectId, queryContent).execute();
+        if (response != null && response.getJobComplete()) {
+          return response;
+        }
       } catch (IOException e) {
         // ignore and retry
         LOG.warn("Ignore the error and retry the query.");
         lastException = e;
       }
     } while(BackOffUtils.next(sleeper, backOff));
-    throw new IOException(
-        String.format(
-            "Unable to get BigQuery response after retrying %d times", MAX_QUERY_RETRIES),
-        lastException);
+
+    if (lastException != null) {
+      throw new IOException(
+          String.format(
+              "Unable to get BigQuery response after retrying %d times", MAX_QUERY_RETRIES),
+          lastException);
+    }
+    return response;
```

Comment:
describeMismatchSafely is default to be used for describing why matcher not accept the item. "The description will be part of a larger description of why a matching failed". However, since we didn't directly use the given item for comparison, this method may be responsible for describing why data from Bigquery doesn't match. 

The main reason I handle 2) and 3) together in describeMismatchSafely is to avoid using exception to terminate test run. However, as you said, it makes sense that 1) and 2) can handle together here with more concise message. Does it make sense to handle 3) in describeMismatchSafely?